Framework/Reflectometry/src/ReflectometryReductionOneAuto3.cpp

```diff
@@ -842,144 +853,163 @@ bool ReflectometryReductionOneAuto3::processGroups() {
+void ReflectometryReductionOneAuto3::setOutputWorkspaces(
+    std::vector<std::string> IvsQGroup,
+    std::vector<std::string> IvsQBinnedGroup,
+    std::vector<std::string> IvsLamGroup, WorkspaceNames const &outputNames) {
   Algorithm_sptr groupAlg = createChildAlgorithm("GroupWorkspaces");
   groupAlg->setChild(false);
   groupAlg->setRethrows(true);
-  if (!IvsLamGroup.empty()) {
+  if (anyWorkspaceInListExists(IvsQGroup)) {
+    groupAlg->setProperty("InputWorkspaces", IvsQGroup);
+    groupAlg->setProperty("OutputWorkspace", outputNames.iVsQ);
+    groupAlg->execute();
+  }
+  if (anyWorkspaceInListExists(IvsQBinnedGroup)) {
+    groupAlg->setProperty("InputWorkspaces", IvsQBinnedGroup);
+    groupAlg->setProperty("OutputWorkspace", outputNames.iVsQBinned);
+    groupAlg->execute();
+  }
+  if (anyWorkspaceInListExists(IvsLamGroup)) {
     groupAlg->setProperty("InputWorkspaces", IvsLamGroup);
-    groupAlg->setProperty("OutputWorkspace", output.iVsLam);
+    groupAlg->setProperty("OutputWorkspace", outputNames.iVsLam);
     groupAlg->execute();
   }
-  groupAlg->setProperty("InputWorkspaces", IvsQBinnedGroup);
-  groupAlg->setProperty("OutputWorkspace", output.iVsQBinned);
-  groupAlg->execute();
-  groupAlg->setProperty("InputWorkspaces", IvsQGroup);
-  groupAlg->setProperty("OutputWorkspace", output.iVsQ);
-  groupAlg->execute();
 
-  // Set other properties so they can be updated in the Reflectometry interface
-  setPropertyValue("ThetaIn", alg->getPropertyValue("ThetaIn"));
-  setPropertyValue("MomentumTransferMin",
-                   alg->getPropertyValue("MomentumTransferMin"));
-  setPropertyValue("MomentumTransferMax",
-                   alg->getPropertyValue("MomentumTransferMax"));
-  setPropertyValue("MomentumTransferStep",
-                   alg->getPropertyValue("MomentumTransferStep"));
-  setPropertyValue("ScaleFactor", alg->getPropertyValue("ScaleFactor"));
+  setPropertyValue("OutputWorkspace", outputNames.iVsQ);
+  setPropertyValue("OutputWorkspaceBinned", outputNames.iVsQBinned);
+  setPropertyValue("OutputWorkspaceWavelength", outputNames.iVsLam);
+}
+
+void ReflectometryReductionOneAuto3::setOutputProperties(
+    OutputProperties const &outputProperties) {
+  setProperty("ThetaIn", outputProperties.thetaIn);
+  setProperty("MomentumTransferMin", outputProperties.qMin);
+  setProperty("MomentumTransferMax", outputProperties.qMax);
+  setProperty("MomentumTransferStep", outputProperties.qStep);
+  setProperty("ScaleFactor", outputProperties.scale);
+}
 
-  setOutputWorkspaces(output, IvsLamGroup, IvsQBinnedGroup, IvsQGroup);
+/** Process groups. Groups are processed differently depending on transmission
+ * runs and polarization analysis.
+ */
+bool ReflectometryReductionOneAuto3::processGroups() {
+  // this algorithm effectively behaves as MultiPeriodGroupAlgorithm
+  m_usingBaseProcessGroups = true;
 
+  // Get our input workspace group
+  auto group = AnalysisDataService::Instance().retrieveWS<WorkspaceGroup>(
```

Comment:
This adds another ~40 LOC to this method bringing it up to ~90 lines for the single method, making it difficult to keep track of all the way through. Could we further chunk down the two big loops....

Reply:
Well it was originally 166 LOC so I've already removed a lot! But yes it could be refactored further.